### PR TITLE
Fix IDENTITY-5924 SAML2 SSO Carbon Authenticator cannot handle multiple AttributeValue elements in role claim

### PR DESCRIPTION
**Pull request by madurangasiriwardena**

### CODE REVIEW

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java [hunk 1]
```diff
@@ -790,23 +790,26 @@ private String generatePassword(String username) {
                 for (Attribute attribute : attributesList) {
                     String attributeName = attribute.getName();
                     if (attributeName != null && roleClaim.equals(attributeName)) {
-                        // Assumes role claim appear only once
-                        Element value = attribute.getAttributeValues().get(0).getDOM();
-                        String attributeValue = value.getTextContent();
+                        List<XMLObject> attributeValues = attribute.getAttributeValues();
+                        String attributeSeparator = getAttributeSeperator();
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java]: shall we use multipleAttributeValueSeperator as the variable name?

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    private static Assertion generateAssertion() throws ConfigurationException {
+
+        // Builder Attributes
+        SAMLObjectBuilder attrStatementBuilder = (SAMLObjectBuilder) getSAMLBuilder().getBuilder
+                (AttributeStatement.DEFAULT_ELEMENT_NAME);
+        AttributeStatement attrStatement = (AttributeStatement) attrStatementBuilder.buildObject();
+
+        Attribute attrFirstName = buildStringAttribute("http://wso2.org/claims/firstName", new String[]{"John"});
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java]: Shall we use constants for claim URIs?

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java [hunk 3]
```diff
@@ -790,23 +790,30 @@ private String generatePassword(String username) {
...
-
-                        if (log.isDebugEnabled()) {
-                            log.debug("AttributeName : " + attributeName + ", AttributeValue : " + attributeValue);
-                        }
-
-                        roles = attributeValue.split(getAttributeSeperator());
-                        if (log.isDebugEnabled()) {
-                            log.debug("Role list : " + Arrays.toString(roles));
+                        List<XMLObject> attributeValues = attribute.getAttributeValues();
+                        if (attributeValues != null && attributeValues.size() == 1) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java]: Would not it better to combine the attributeValues.size() == 1 and attributeValues.size() > 1) logic together? does not one element contain "multiAttributeSeparator" separated attributes too?

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java [hunk 4]
```diff
@@ -790,23 +790,30 @@ private String generatePassword(String username) {
...
+                            String attributeValueString = ((XSStringImpl) attributeValues.get(0)).getValue();
+                            String multiAttributeSeparator = getAttributeSeperator();
+                            String[] attributeValuesArray = attributeValueString.split(multiAttributeSeparator);
+                            roles.addAll(Arrays.asList(attributeValuesArray));
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java]: Better to have debug log showing the split attributes.

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java [hunk 5]
```diff
@@ -790,23 +790,30 @@ private String generatePassword(String username) {
...
+                        } else if (attributeValues != null && attributeValues.size() > 1) {
+                            for (XMLObject attributeValue : attributeValues) {
+                                String attributeValueString = ((XSStringImpl) attributeValue).getValue();
+                                if (log.isDebugEnabled()) {
+                                    log.debug("AttributeName : " + attributeName + ", AttributeValue : " +
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java]: Can improve log, "Adding attributes for Assertion: "+assertion+" AttributeName : " + attributeName + ...

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java [hunk 6]
```diff
@@ -790,23 +790,30 @@ private String generatePassword(String username) {
...
+                                roles.add(attributeValueString);
+                            }
                         }
                     }
                 }
             }
         }
-        return roles;
+        if (log.isDebugEnabled()) {
+            log.debug("Role list : " + roles);
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java]: Better to improve log 
Role list found for assertion: "+assertion+ ", roles: " + roles

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java (hunk 3 above)
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java]: The idea is role list can either come as a multiAttributeSeparator separated list in a single AttributeValue element or separately in multiple AttributeValue elements. So I don't think we need to consider having multiple roles in single AttributeValue if there are multiple AttributeValue elements

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            DefaultBootstrap.bootstrap();
+            builderFactory = Configuration.getBuilderFactory();
+        }
+
+        return builderFactory;
+    }
+
+    private static Assertion generateAssertion(boolean userMultiAttributeSeparator) throws ConfigurationException {
+
+        // Builder Attributes
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java]: Typo in the comment

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            // OpenSAML 2.3
+            System.setProperty("org.opensaml.httpclient.https.disableHostnameVerification", "true");
+            DefaultBootstrap.bootstrap();
+            builderFactory = Configuration.getBuilderFactory();
+        }
+
+        return builderFactory;
+    }
+
+    private static Assertion generateAssertion(boolean userMultiAttributeSeparator) throws ConfigurationException {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java]: typo userMultiAttributeSeparator

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                .DEFAULT_ELEMENT_NAME);
+        Assertion assertion = (Assertion) assertionBuilder.buildObject();
+        assertion.setVersion(SAMLVersion.VERSION_20);
+
+        assertion.getAttributeStatements().add(attrStatement);
+
+        return assertion;
+    }
+
+    private static Attribute buildStringAttribute(String name, String[] values, boolean userMultiAttributeSeparator)
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/src/test/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticatorTest.java]: typo userMultiAttributeSeparator

File: components/org.wso2.carbon.identity.authenticator.saml2.sso/pom.xml [hunk 10]
```diff
@@ -63,11 +113,41 @@
...
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>
         <plugins>
             <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>2.3</version>
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.authenticator.saml2.sso/pom.xml]: Add a property for the version.

**Pull request merged**